datasets/flwr_datasets/partitioner/size_partitioner.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2023 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""DataQuantityPartitioner class."""
+from typing import Callable, Dict, List, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class SizePartitioner(Partitioner):
+    """Base class for the deterministic size partitioning schemes based on the cid.
+
+    The client with cid has the following relationship regarding the number of samples.
+    cid_to_size_fnc(cid) ~ number of samples for cid
+
+    If the function doesn't transform the cid it's a linear correlations of the number
+    of sample for cid and the value of cid. For instance, if the IDs range from 1 to M,
+    client with ID 1 gets 1 unit of data, client 2 gets 2 units, and so on, up to client
+     M which gets M units.
+
+    Note that size corresponding to the cid is deterministic, yet in case of different
+    dataset shuffling the assignment of samples to cid will vary.
+
+    Parameters
+    ----------
+    num_partitions: int
+        The total number of partitions that the data will be divided into.
+    cid_to_size_fnc: Callable
+        Function that defines the relationship between cid and the number of samples.
+    """
+
+    def __init__(
+        self,
+        num_partitions: int,
+        cid_to_size_fnc: Callable,  # type: ignore[type-arg]
+    ) -> None:
+        super().__init__()
+        if num_partitions <= 0:
+            raise ValueError("The number of partitions must be greater than zero.")
+        self._num_partitions = num_partitions
+        self._cid_to_size_fnc = cid_to_size_fnc
+
+        self._cid_to_size: Dict[int, int] = {}
+        self._cid_to_indices: Dict[int, List[int]] = {}
+        # A flag to perform only a single compute to determine the indices
+        self._cid_to_indices_determined = False
+
+    def load_partition(self, idx: int) -> datasets.Dataset:
+        """Load a single partition based on the partition index.
+
+        For this partitioner the number of samples is dependent on the partition idx.
+
+        Parameters
+        ----------
+        idx: int
```

Comment:
The NumPy convention is to have a space before the `:`, please update other comments in this PR as well:
```suggestion
        idx : int
```

Reply:
Done. Unfortunately, I was not aware of that. It's like that all over the project. I'll bridge that gap for the rest of the code in the next PR